### PR TITLE
Switch from containers/image to go-containerregistry

### PR DESCRIPTION
**Pull request by nkubala**
This removes the dependency on containers/image in favor of our new go-containerregistry library.

Filesystem diffs are now done by extracting the flattened fs from the image.

Caching has been temporarily removed while I design a new solution, since the layer caching will no longer work. We can probably just naively cache the fs tarball for now.

The RPM differ needed to be rewritten to load the image into the docker daemon before running commands in a container. https://github.com/google/go-containerregistry/pull/114 needs to be merged before this will compile, but should work correctly once we update the deps to pick up that commit.

### CODE REVIEW

**Comment by vrothberg** [comment]
@nkubala Out of curiosity: What was the reasoning behind writing the `go-containerregistry` library? I am asking as I consider the `containers/*` libraries to be quite feature rich and upstream is nice to work with.

**Comment by nkubala** [comment]
@vrothberg Google has had https://github.com/google/containerregistry, a registry library written in Python, for over a year now. Since most of our tools that we're working on these days are written in Go, we decided to port the library over. containers/image is definitely nice, but maintaining our own library gives us more flexibility and control over design decisions.

File: cmd/root.go [hunk 1]
```diff
@@ -121,67 +124,67 @@ func checkIfValidAnalyzer(_ []string) error {
...
-	}
-
-	if pkgutil.IsTar(image) {
-		return &pkgutil.TarPrepper{
-			Source: filepath.Clean(image),
-			Client: cli,
-		}, nil
+func getImageForName(imageName string) (pkgutil.Image, error) {
+	if !pkgutil.IsTar(imageName) && !pkgutil.HasTag(imageName) {
+		imageName = imageName + pkgutil.LatestTag
```
Comment (dlorenc) [inline cmd/root.go]: Why do you have to add latest? WeakValidation should add that below.

File: cmd/root.go [hunk 2]
```diff
@@ -121,67 +124,67 @@ func checkIfValidAnalyzer(_ []string) error {
...
+		imageName = strings.Replace(imageName, RemotePrefix, "", -1)
+		ref, err := name.ParseReference(imageName, name.WeakValidation)
+		if err != nil {
+			return pkgutil.Image{}, err
+		}
+		auth, err := authn.DefaultKeychain.Resolve(ref.Context().Registry)
+		if err != nil {
+			return pkgutil.Image{}, err
+		}
+		// logrus.Infof("retrieving remote image: %v", ref)
```
Comment (dlorenc) [inline cmd/root.go]: uncomment?

File: differs/metadata_diff.go [hunk 3]
```diff
@@ -53,14 +59,50 @@ func (a MetadataAnalyzer) Analyze(image pkgutil.Image) (util.Result, error) {
...
-func getMetadataList(image pkgutil.Image) []string {
-	return image.Config.Config.AsList()
+func getMetadataList(image pkgutil.Image) ([]string, error) {
+	configFile, err := image.Image.ConfigFile()
+	if err != nil {
+		return nil, err
+	}
+	c := configFile.Config
+	return []string{
+		fmt.Sprintf("Hostname: %s", c.Hostname),
```
Comment (dlorenc) [inline differs/metadata_diff.go]: Do these need newlines?

File: differs/metadata_diff.go [hunk 4]
```diff
@@ -53,14 +59,50 @@ func (a MetadataAnalyzer) Analyze(image pkgutil.Image) (util.Result, error) {
...
+		fmt.Sprintf("ExposedPorts: %v", pkgutil.SortMap(c.ExposedPorts)),
+		fmt.Sprintf("Tty: %t", c.Tty),
+		fmt.Sprintf("OpenStdin: %t", c.OpenStdin),
+		fmt.Sprintf("StdinOnce: %t", c.StdinOnce),
+		fmt.Sprintf("Env: %s", strings.Join(c.Env, ",")),
+		fmt.Sprintf("Cmd: %s", strings.Join(c.Cmd, ",")),
+		fmt.Sprintf("ArgsEscaped: %t", c.ArgsEscaped),
+		fmt.Sprintf("Image: %s", c.Image),
+		fmt.Sprintf("Volumes: %v", pkgutil.SortMap(c.Volumes)),
+		// fmt.Sprintf("Workdir: %s", c.Workdir),
```
Comment (dlorenc) [inline differs/metadata_diff.go]: uncomment?

File: cmd/root.go (hunk 1 above)
Comment (nkubala) [reply cmd/root.go]: ah you're right, removed

File: cmd/root.go (hunk 2 above)
Comment (nkubala) [reply cmd/root.go]: done

File: differs/metadata_diff.go (hunk 3 above)
Comment (nkubala) [reply differs/metadata_diff.go]: they don't, this is a list of strings that gets thrown into a generic result struct and outputted separately

File: differs/metadata_diff.go (hunk 4 above)
Comment (nkubala) [reply differs/metadata_diff.go]: ah yeah thanks, fixed workdir and removed stoptimeout since it's not part of the returned config

**Comment by mattmoor** [comment]
@nkubala nit: We've had it for about 4 years, but most of that internally :)

**Pull request merged**